packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+    api,
+    credentialId,
+    blockNumber
+  )
+
+  if (extrinsic === null) {
+    throw new SDKErrors.PublicCredentialError(
+      `The block number as specified in the provided credential entry (${blockNumber}) does not have any extrinsic that includes a credential creation.`
+    )
+  }
+
+  if (!api.tx.did.submitDidCall.is(extrinsic)) {
+    throw new SDKErrors.PublicCredentialError(
+      'Extrinsic should be a did.submitDidCall extrinsic'
+    )
+  }
+
+  const [extrinsicCalls, extrinsicDidOrigin] = [
+    flattenCalls(api, extrinsic.args[0].call),
+    didFromChain(extrinsic.args[0].did),
+  ]
+
+  let credentialInput: IPublicCredential | undefined
+
+  // Iterate over the calls in the extrinsic to find the right one, and re-create the issued public credential.
+  // If more than a call are present, it always consider the last one as the valid one.
+  extrinsicCalls.forEach((call) => {
+    if (api.tx.publicCredentials.add.is(call)) {
+      const credentialCallArgument = call.args[0]
+      const reconstructedCredentialInput = credentialInputFromChain(
+        credentialCallArgument
+      )
+      const reconstructedId = computeId(
+        reconstructedCredentialInput,
+        extrinsicDidOrigin
+      )
+      if (reconstructedId === credentialId) {
+        credentialInput = {
+          ...reconstructedCredentialInput,
+          attester: extrinsicDidOrigin,
+          id: reconstructedId,
+          blockNumber,
+          revoked: revoked.toPrimitive(),
+        }
+      }
+    }
+  })
+  if (!credentialInput) {
+    throw new SDKErrors.PublicCredentialError(
+      'Block should always contain the full credential, eventually.'
+    )
+  }
+  return credentialInput
+}
+
+/**
+ * Decodes the public credential details returned by `api.call.publicCredential.getCredentials()`.
+ *
+ * This is the **only** secure way for users to retrieve and verify all the credentials issued to a given [[AssetDidUri]].
+ * Hence, calling `api.call.publicCredentials.getCredentials(asset_id)` and then passing the result to this function is the only way to trust that the credentials for a given AssetDID are valid.
+ *
+ * @param publicCredentialEntries The raw public credential details from blockchain.
+ * @returns An array of [[IPublicCredential]] as the result of combining the on-chain information and the information present in the tx history. If the result is an error, it maps it to the right error type.
+ */
+export async function credentialsFromChain(
+  publicCredentialEntries: Result<
+    Vec<ITuple<[Hash, PublicCredentialsCredentialsCredentialEntry]>>,
+    PublicCredentialError
+  >
+): Promise<IPublicCredential[]> {
+  if (publicCredentialEntries.isErr) {
+    throw new Error(publicCredentialEntries.asErr.toString())
+  }
+
+  const api = ConfigService.get('api')
+  const formattedCredentials: Array<
+    [HexString, Option<PublicCredentialsCredentialsCredentialEntry>]
```

Comment:
Doesn’t TS infer this type automatically?

Reply:
```
Argument of type '`0x${string}` | Option<PublicCredentialsCredentialsCredentialEntry>' is not assignable to parameter of type '`0x${string}`'.
  Type 'Option<PublicCredentialsCredentialsCredentialEntry>' is not assignable to type '`0x${string}`'.
```

Reply:
Apparently returning a tuple makes all elements of the tuple type unions...

Reply:
I see… With that knowledge, it looks like creating an intermediate data structure harms more than it helps. Would it become better when calling `credentialFromChain` directly in the same loop?

Reply:
I personally don't think this is any clearer, but I followed your suggestion regardless: https://github.com/KILTprotocol/sdk-js/pull/677/commits/589c8bc7d8f0fbdf240c2824a70adb34102dbfb0#diff-3228260bd7a8a1c0fb1e0b94efbc0b692518737b171dda86329b57db740703bbR208.

Reply:
I think it has less noise this way :+1: